### PR TITLE
Add enum_iterator type

### PR DESCRIPTION
**Pull request by asklar**
This change adds an `enum_iterator` type which allows iterating in a natural way over `IEnum`-like COM interfaces, like `IEnumIDList`, `IDiaEnumSymbols`, etc.

bonus track: also fixes a few unrelated typos

### CODE REVIEW

File: include/wil/enum_iterator.h [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+            using ArgTypes = std::tuple<Args...>;
+            static constexpr std::size_t ArgCount = sizeof...(Args);
+            template<std::size_t N>
+            using NthArg = std::tuple_element_t<N, ArgTypes>;
+        };
+
+        template<typename F> struct FunctionTraits;
+
+        template<typename T, typename Arg>
+        struct FunctionTraits<HRESULT(__stdcall T::*)(ULONG, Arg**, ULONG*)>
```
Comment (asklar) [inline include/wil/enum_iterator.h:41]:
```suggestion
        template<typename T, typename Arg>
        using Next_t = HRESULT(__stdcall T::*)(ULONG, Arg**, ULONG*);

        template<typename T, typename Arg>
        struct FunctionTraits<Next_t<T, Arg>>
```

File: include/wil/enum_iterator.h [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+//*********************************************************
+#ifndef __WIL_ENUM_ITERATOR
+#define __WIL_ENUM_ITERATOR
+
+#include "com.h"
+#include "result_macros.h"
+#include <type_traits>
+#include <tuple>
+
+
```
Comment (ChrisGuzak) [inline include/wil/enum_iterator.h:19]: nit, remove extra line

File: include/wil/enum_iterator.h [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+#include <type_traits>
+#include <tuple>
+#include <cstdio>
+
+namespace wil {
```
Comment (ChrisGuzak) [inline include/wil/enum_iterator.h:24]: nits, curly braces on their own lines

File: include/wil/enum_iterator.h [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+            (*this)++;
+        }
+        /// <summary>
+        /// The type of the out parameter in Next(...) that corresponds to the element type being enumerated.
+        /// e.g. PITEMID_CHILD, IDiaSymbol, etc.
+        /// </summary>
+        using TElem = Details::NextArgType<TEnum>;
+        ~enum_iterator() = default;
+
+        auto& operator++(int) { return this->operator++(); }
```
Comment (ChrisGuzak) [inline include/wil/enum_iterator.h:73]: I probably have this wrong, but isn't the `int` the # of items to increment? can you pass it to `this->operator++'?

`auto& operator++(int num) { return this->operator++(num); }`

File: include/wil/enum_iterator.h [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+        auto& operator++() {
+            TElem* pChild{ nullptr };
+            ULONG celt = 0;
+            auto hr = m_pEnum->Next(1, &pChild, &celt);
+            if (hr == S_OK && celt == 1) {
```
Comment (ChrisGuzak) [inline include/wil/enum_iterator.h:78]: nit, parens around sub-expressions

File: include/wil/enum_iterator.h [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+            auto other = *this;
+            for (int i = 0; i < v; i++) {
+                other++;
+            }
+            return other;
+        }
+
+        bool operator==(const enum_iterator& o) const noexcept {
+            if (m_end ^ o.m_end) return false;
+            if (m_end == o.m_end && m_end == true) return true;
```
Comment (ChrisGuzak) [inline include/wil/enum_iterator.h:103]: nits, sub-expression parens

**Review by ChrisGuzak — CHANGES_REQUESTED** [review]
all my comments were minor nits...

File: include/wil/enum_iterator.h (hunk 4 above)
Comment (asklar) [reply include/wil/enum_iterator.h:73]: It's the postincrement operator (the one without the int is the preincrement operator), i.e. `x++` vs. `++x`.
see [Increment and Decrement Operator Overloading (C++)](https://docs.microsoft.com/cpp/cpp/increment-and-decrement-operator-overloading-cpp)

File: include/wil/enum_iterator.h [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+        /// The type of the out parameter in Next(...) that corresponds to the element type being enumerated.
+        /// e.g. PITEMID_CHILD, IDiaSymbol, etc.
+        /// </summary>
+        using TElem = Details::NextArgType<TEnum>;
+        ~enum_iterator() = default;
+
+        auto& operator++(int) { return this->operator++(); }
+        auto& operator++()
+        {
+            TElem* pChild{ nullptr };
```
Comment (jonwis) [inline include/wil/enum_iterator.h:78]: This appears to leak TElem and then `m_current` as the iterator moves down the collection.

File: include/wil/enum_iterator.h [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+        static constexpr auto end() noexcept
+        {
+            return enum_iterator{ end_tag{} };
+        }
+    protected:
+        struct end_tag {};
+        enum_iterator(end_tag) : m_end(true) {}
+        enum_iterator() {};
+        wil::com_ptr<TEnum> m_pEnum{ nullptr };
+        TElem* m_current{ nullptr };
```
Comment (jonwis) [inline include/wil/enum_iterator.h:127]: What about collections of scalars?

**Review by jonwis — CHANGES_REQUESTED** [review]

File: include/wil/enum_iterator.h (hunk 7 above)
Comment (asklar) [reply include/wil/enum_iterator.h]: Good catch! I had originally used a com_ptr<TElem> but that didn't work for enumerating non-COM objects, obviously, and I ended up leaking it; updated the code to conditionally use com_ptr based on whether the type has an AddRef method in it 🤓

File: include/wil/enum_iterator.h (hunk 8 above)
Comment (asklar) [reply include/wil/enum_iterator.h]: yeah I need to think about it more; esp. other non-COM types that still need to be resource-managed (like ITEMID_CHILD needs to be ComTaskFree'd)

**Review by ChrisGuzak — APPROVED** [review]

File: include/wil/enum_iterator.h [hunk 9]
```diff
@@ -0,0 +1,225 @@
...
+            {
+                return m_pEnum->Next(1, m_current.put(), celt);
+            }
+            else if constexpr (is_cotaskmem_ptr)
+            {
+                TElemRaw* rawPtr{ nullptr };
+                auto hr = m_pEnum->Next(1, &rawPtr, celt);
+                m_current.reset(rawPtr, wil::cotaskmem_deleter{});
+                return hr;
+            }
```
Comment (jonwis) [inline include/wil/enum_iterator.h:148]: This should probably have an `else { static_assert(false); }` in it?

File: include/wil/enum_iterator.h [hunk 10]
```diff
@@ -0,0 +1,225 @@
...
+        }
+
+
+    public:
+        auto& operator++(int) { return this->operator++(); }
+        auto& operator++()
+        {
+            TElem pChild{ };
```
Comment (jonwis) [inline include/wil/enum_iterator.h:156]: `pChild` is unused here.

**Review by jonwis — APPROVED** [review]

**Comment by dunhor** [comment]
@asklar should we close this now?

**Pull request closed without merging**